video/agon_ps2.h

```diff
@@ -0,0 +1,371 @@
+		*modifiers = 0;
+		*vk = 0;
+		*down = 0;
+		return true;
+	}
+	#endif
+	
+	if (kb->getNextVirtualKey(&item, 0)) {
+		if (item.down) {
+			switch (item.vk) {
+				case fabgl::VK_LEFT:
+					_keycode = 0x08;
+					break;
+				case fabgl::VK_TAB:
+					_keycode = 0x09;
+					break;
+				case fabgl::VK_RIGHT:
+					_keycode = 0x15;
+					break;
+				case fabgl::VK_DOWN:
+					_keycode = 0x0A;
+					break;
+				case fabgl::VK_UP:
+					_keycode = 0x0B;
+					break;
+				case fabgl::VK_BACKSPACE:
+					_keycode = 0x7F;
+					break;
+				default:
+					_keycode = item.ASCII;	
+					break;
+			}
+			// Pack the modifiers into a byte
+			//
+			_modifiers = 
+				item.CTRL		<< 0 |
+				item.SHIFT		<< 1 |
+				item.LALT		<< 2 |
+				item.RALT		<< 3 |
+				item.CAPSLOCK	<< 4 |
+				item.NUMLOCK	<< 5 |
+				item.SCROLLLOCK << 6 |
+				item.GUI		<< 7
+			;
+		}
+		*keycode = _keycode;
+		*modifiers = _modifiers;
+		*vk = item.vk;
+		*down = item.down;
+
+		return true;
+	}
+
+	return false;
+}
+
+// Simpler keyboard read for CP/M Terminal Mode
+//
+bool getKeyboardKey(uint8_t *ascii) {
+	auto kb = getKeyboard();
+	fabgl::VirtualKeyItem item;
+
+	// Read the keyboard and transmit to the Z80
+	//
+	if (kb->getNextVirtualKey(&item, 0)) {
+		if (item.down) {
+			*ascii = item.ASCII;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+// Wait for shift key to be released, then pressed (used for paged mode)
+// 
+bool wait_shiftkey(uint8_t *ascii, uint8_t* vk, uint8_t* down) {
+	auto kb = getKeyboard();
+	fabgl::VirtualKeyItem item;
+
+	// Wait for shift to be released
+	//
+	do {
+		kb->getNextVirtualKey(&item, 0);
+	} while (item.SHIFT);
+
+	// And pressed again
+	//
+	do {
+		kb->getNextVirtualKey(&item, 0);
+		*ascii = item.ASCII;
+		*vk = item.vk;
+		*down = item.down;
+		if (item.ASCII == 27) {	// Check for ESC
+			return false;
+		}
+	} while (!item.SHIFT);
+
+	return true;
+}
+
+void getKeyboardState(uint16_t *repeatDelay, uint16_t *repeatRate, uint8_t *ledState) {
+	bool numLock;
+	bool capsLock;
+	bool scrollLock;
+	getKeyboard()->getLEDs(&numLock, &capsLock, &scrollLock);
+	*ledState = scrollLock | (capsLock << 1) | (numLock << 2);
+    *repeatDelay = kbRepeatDelay;
+    *repeatRate = kbRepeatRate;
+}
+
+void setKeyboardState(uint16_t delay, uint16_t rate, uint8_t ledState) {
+	auto kb = getKeyboard();
+	if (delay >= 250 && delay <= 1000) kbRepeatDelay = (delay / 250) * 250;	// In steps of 250ms
+	if (rate >=  33 && rate <=  500) kbRepeatRate  = rate;
+	if (ledState != 255) {
+		kb->setLEDs(ledState & 4, ledState & 2, ledState & 1);
+	}
+    kb->setTypematicRateAndDelay(kbRepeatRate, kbRepeatDelay);
+}
+
+bool updateMouseEnabled() {
+	auto mouse = getMouse();
+	if (!mouse) {
+		return false;
+	}
+	mouseEnabled = mouse->isMouseAvailable();
+	return mouseEnabled;
+}
+
+bool enableMouse() {
+	if (mouseEnabled) {
+		return true;
+	}
+	auto mouse = getMouse();
+	if (!mouse) {
+		return false;
+	}
+	mouse->resumePort();
+	return updateMouseEnabled();
```

Comment:
this ensures that the mouse will only be enabled if a mouse is actually available.

in the current implementation if you turn on a machine that does _not_ have a mouse connected, then calling "enable" will fail, and the mouse will not enable even if you subsequently plug a mouse in.  to get the mouse to be recognised you either need to call the `reset` command (`VDU 23,0,&89,2`) _or_ press the reset button.  this is an intentional choice

in principle this code _could_ attempt to reset the mouse system and connect to a mouse if no mouse had previously been detected, but that is very costly and takes a while (up to a couple of seconds), jamming up the VDP in the process.  as many Agon Light machines will not have a mouse port, and thus won't have a mouse connected, it's better to make this "enable" fail quickly.  similarly most users that _do_ have a mouse will keep it plugged in